chain/jsonrpc/src/lib.rs

```diff
@@ -958,6 +971,45 @@ impl JsonRpcHandler {
     }
 }
 
+#[cfg(feature = "sandbox")]
+impl JsonRpcHandler {
+    async fn sandbox_patch_state(
+        &self,
+        patch_state_request: near_jsonrpc_primitives::types::sandbox::RpcSandboxPatchStateRequest,
+    ) -> Result<
+        near_jsonrpc_primitives::types::sandbox::RpcSandboxPatchStateResponse,
+        near_jsonrpc_primitives::types::sandbox::RpcSandboxPatchStateError,
+    > {
+        self.client_addr
+            .send(NetworkClientMessages::Sandbox(NetworkSandboxMessage::SandboxPatchState(
+                patch_state_request.records,
+            )))
+            .await?;
+
+        timeout(self.polling_config.polling_timeout, async {
+            loop {
+                let patch_state_finished = self
+                    .client_addr
+                    .send(NetworkClientMessages::Sandbox(
+                        NetworkSandboxMessage::SandboxPatchStateStatus {},
+                    ))
+                    .await;
+                if let Ok(NetworkClientResponses::SandboxResult(
+                    SandboxResponse::SandboxPatchStateFinished(true),
+                )) = patch_state_finished
+                {
+                    break;
+                }
+                let _ = sleep(self.polling_config.polling_interval).await;
+            }
+        })
+        .await
+        .expect("patch state should happen at next block, never timeout");
```

Comment:
Hm, this gives me a pause. The problem with poling loops is that they are slow -- you generally add at least half of the polling interval to the waiting. This tends to kill test performance --- if each test has to wait 100 milliseconds, but you have a thousands of tests, your test suite spends over a minute waiting! 

This code is going to be hit by contract authors who write the tests. That is, it's not our test suite which is going to be slow (which we can live with), but the test suite of developers building on top of NEAR. 

Is there some existing issue which tracks the reason we do polling loops here? My understanding is that this all happens withing a single process/address spaces, so I don't see an architectural reason for us to poll via a loop.

Reply:
No, this not the test noticably, patch_state have to happen in a block, which takes at least a second (block time). user expected to wait 2-3 seconds for block confirmation even this is without pool.
Also it's not possible to synchronously wait for chain to confirm "this state has been patched", current architecture rpc server can send NetworkMessage to Chain, chain doing the job (process blocks, patch states), but chain cannot notify back the operation is done if operation is async, such as submit a txn or patch state, both are pooled and effective later when block is produced. So, rpc server has to send another kind of message to Chain to pool this status.